bb/pm-19497/replace default reset button to enable it in more browsers

bensbits91 opened the pull request:
## 🎟️ Tracking

<!-- Paste the link to the Jira or GitHub issue or otherwise describe / point to where this change is coming from. -->

GitHub Issue: https://github.com/bitwarden/clients/issues/13982

Jira ticket number: PM-19497

## 📔 Objective

<!-- Describe what the purpose of this PR is, for example what bug you're fixing or new feature you're adding. -->

- Update the template (search.component.html) to add a new reset button:
    - Wrap the existing label and input in a form element, which is required for our new custom reset button to work properly.
    - Add event bindings to the input for `focus`, `blur`, `mouseenter` and `mouseleave` to facilitate emulating the hiding/showing of the default reset button.
    - Add the new reset button, including:
        - An `*ngIf="searchText"` directive and event bindings to enable hiding our new button when the input is empty.
        - Class bindings to reproduce our old reset button style.
        - The (click) event handler that triggers clearing the input.
- Update the class to:
    - Track the input's focus/blur and hover states (to facilitate emulating the hiding/showing of the default reset button).
    - Update the model in the `onChange` lifecycle hook (i.e. `this.searchText = searchText;`), to enable hiding our new reset button when the input is empty.
    - Add the `clearSearch` method that clears the input, which is called by clicking our new reset button.
- Update CSS (search.component.css) to hide the default reset button, which only appears in some browsers (e.g. Chrome).

## 📸 Screenshots

<!-- Required for any UI changes; delete if not applicable. Use fixed width images for better display. -->

![image](https://github.com/user-attachments/assets/41967e97-75ac-4524-8a00-63f9800fcce3)

![image](https://github.com/user-attachments/assets/efb58c6a-2376-42b3-924d-d5d2ff7e42cf)



## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

vleague2 reviewed (changes requested):
Thanks for the PR! This is definitely an issue we've had multiple reports of, so it would be great to get it resolved.

I've left some comments on a couple of things that we would normally be looking at when updating a component library component:
- Accessibility concerns
- Common syntax/implementation style
- Reusability

The last one, in this PR, has manifested as taking the opportunity to explore a way to reuse an existing component to accomplish what we're doing with this one (the comment about the `bit-form-field`). Let me know your thoughts, and if it's something you would be interested in working on. (Totally fine if not!)

vleague2 commented:
@bensbits91 Absolutely, no rush! Have fun camping!

bensbits91 commented on libs/components/src/search/search.component.html:
```diff
@@ -1,23 +1,39 @@
-    [disabled]="disabled"
-    [attr.autocomplete]="autocomplete"
-  />
+  <form class="tw-relative tw-flex tw-items-center tw-w-full">
```
>❓ I know you said the form is required in the PR description, but can you give a little more info on why?

To my knowledge, an input with type = reset is intended to clear the inputs that are contained in the same form. I tried without wrapping in a form, and the reset button didn't do anything. Let me know if there's a better solution (I'm sure we could devise something if we want to avoid adding the parent form element).


>⚠️ Since this wraps a search element, it should have role="search" on it for accessibility ([source](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/search_role))

Thanks! I added role="search" to the form element.


>💭 I also wonder if the sr-only label should be inside of the form element now that it exists

Totally! I moved it inside the form, next to the other label element.

bensbits91 commented on libs/components/src/search/search.component.html:
```diff
@@ -1,23 +1,39 @@
+      (mouseleave)="isInputHovered = false"
+      [disabled]="disabled"
+      [attr.autocomplete]="autocomplete"
+    />
```
I'd love to explore/tackle this with you 👍

bensbits91 commented on libs/components/src/search/search.component.ts:
```diff
@@ -43,6 +44,10 @@ export class SearchComponent implements ControlValueAccessor, FocusableElement {
+  // Optional: Track focus and hover state of the input to emulate the hiding/showing of the native reset button
+  protected isInputFocused = false;
+  protected isInputHovered = false;
+  protected resetHovered = false;
```
@vleague2 I implemented two signals. To emulate the native reset button, I think we need two separate booleans/signals to track focus and hover states. When the input doesn't have focus, the reset button only appears when the input is hovered.

```js
protected isInputFocused = signal(false);
protected isFormHovered = signal(false);

protected showResetButton = computed(() => this.isInputFocused() || this.isFormHovered());
```

I moved the tracking of hover state to the form (because tracking hover states of the input and the button caused conflicts).

```js
<form 
  role="search" 
  (mouseenter)="isFormHovered.set(true)"
  (mouseleave)="isFormHovered.set(false)"
...
>
  ...
  <input
    ...
    (focus)="isInputFocused.set(true)"
    (blur)="isInputFocused.set(false); onTouch()"
    ...
  />
  <button
    *ngIf="searchText && showResetButton()"
    ...
  ></button>
</form>

```

Signals are new to me, so let me know if this is messy and I'll be happy to keep working on it.

Thanks

bensbits91 commented on libs/components/src/search/search.component.ts:
```diff
@@ -53,11 +58,20 @@ export class SearchComponent implements ControlValueAccessor, FocusableElement {
+  clearSearch() {
+    this.searchText = "";
+    if (this.notifyOnChange) {
+      this.notifyOnChange("");
```
It doesn't appear that the reset button calls the input's `onChange` handler. I added temporary console logs to each handler to double-check. The input's `onChange` is called while typing, but not when clicking the reset button. 

Let me know if you'd like me to explore another solution. For example, we can call `onChange` from the button `(click)="clearSearch(); onChange('')"` (and remove the `notifyChange` call from `clearSearch` handler).

bensbits91 commented:
@vleague2 I addressed your comments. I'm happy to keep working on any part of this if I missed the mark or if you have any more suggestions.

Thanks again 😄

bensbits91 commented on libs/components/src/search/search.component.html:
```diff
@@ -1,23 +1,44 @@
-<label class="tw-sr-only" [for]="id">{{ "search" | i18n }}</label>
 <div class="tw-relative tw-flex tw-items-center">
```
❓  I removed the div altogether. You're right that we don't need duplicated classes on the div and form, and without the classes, the div seemed superfluous. Is removing the div okay?

vleague2 replied:
I think that makes sense!

bensbits91 commented on libs/components/src/search/search.component.html:
```diff
@@ -1,23 +1,44 @@
+        'tw-opacity-0': !showResetButton(),
+        'tw-bg-text-muted': showResetButton(),
+      }"
+      class="bw-reset-btn tw-size-6 tw-absolute hover:tw-bg-text-main tw-right-2 tw-z-20 !tw-mb-0 tw-cursor-pointer"
```
✅

bensbits91 commented on libs/components/src/search/search.component.html:
```diff
@@ -1,23 +1,44 @@
+    <label
+      [for]="id"
+      aria-hidden="true"
+      class="tw-absolute tw-left-2 tw-z-20 !tw-mb-0 tw-cursor-text"
```
✅

bensbits91 commented on libs/components/src/search/search.component.html:
```diff
@@ -1,23 +1,44 @@
+      class="tw-pl-9"
+      [ngModel]="searchText"
+      (ngModelChange)="onChange($event)"
+      [ngModelOptions]="{ standalone: true }"
```
✅

bensbits91 commented on libs/components/src/search/search.component.html:
```diff
@@ -1,23 +1,44 @@
+      }"
+      class="bw-reset-btn tw-size-6 tw-absolute hover:tw-bg-text-main tw-right-2 tw-z-20 !tw-mb-0 tw-cursor-pointer"
+      type="reset"
+      [attr.aria-label]="searchText ? ('Reset search' | i18n) : ''"
```
I added "Reset search" to the following files:

- Browser: `apps/browser/src/_locales/en/messages.json`
- Desktop: `apps/desktop/src/locales/en/messages.json`
- Web: `apps/web/src/locales/en/messages.json`

bensbits91 commented on libs/components/src/search/search.component.html:
```diff
@@ -1,23 +1,39 @@
+      (mouseleave)="resetHovered = false"
+      [class.opacity-0]="!isInputFocused && !isInputHovered && !resetHovered"
+      [class.tw-bg-text-muted]="isInputFocused || isInputHovered || resetHovered"
+      class="bw-reset-btn tw-size-6 tw-absolute hover:tw-bg-text-main tw-right-2 tw-z-20 !tw-mb-0 tw-cursor-pointer"
```
I dug a little 😃 Our input uses the `bitinput` directive. In `/libs/components/src/input/input.directive.ts` > `get standaloneInputClasses()`, we have `"focus:tw-z-10"`. So the reset button needs `tw-z-20` to remain visible and interactive when the input has focus. (The label that contains the bwi-search icon also has `tw-z-20`).

vleague2 commented on libs/components/src/search/search.component.ts:
```diff
@@ -30,7 +31,8 @@ let nextId = 0;
     },
   ],
-  imports: [InputModule, ReactiveFormsModule, FormsModule, I18nPipe],
+  standalone: true,
```
Hmm it does seem a little strange that we'd need those, based on where I see those getting used elsewhere in the codebase (Angular forms are not my strong suit either!). What browser are you testing on? I deleted those two lines on the form element and I am still seeing the reset button in Chrome and Firefox. Maybe you can take a video?

bensbits91 replied:
Hmmm, you're right. I removed the 2 `ngForm` attributes (but the input does seem to need the `name` attribute).

bensbits91 commented:
@vleague2 Thanks again for reviewing 😀  I think I addressed your latest comment

vleague2 commented on libs/components/src/search/search.component.ts:
```diff
@@ -30,7 +31,7 @@ let nextId = 0;
-  imports: [InputModule, ReactiveFormsModule, FormsModule, I18nPipe],
+  imports: [InputModule, ReactiveFormsModule, FormsModule, I18nPipe, NgIf, NgClass],
 })
 export class SearchComponent implements ControlValueAccessor, FocusableElement {
```
Nested forms is [not supported by the HTML spec](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/form#technical_summary). I did a spot-check in the code and I don't think we are using the search inside of a form right now, so we should be good. But I do want to make sure devs know about this when using the search component in the app code, so we can make a couple small updates to the Storybook docs.

Update 1:
Add a jsdoc comment to this component (doesn't have to be this exact text, I did not format this in a code editor either lol)
```suggestion
/**
 * Search components must not be nested inside of a form, since they contain their own standalone form for searching.
 */
export class SearchComponent implements ControlValueAccessor, FocusableElement {
```

Update 2:
In `search.mdx`, update line 1 to import "Description" too, then update line 11 to add:
`<Description />`
(you can check an existing mdx page like `icon-button.mdx` for an example) 
That should display the jsdoc comment on the Storybook page.

bensbits91 replied:
I added a comment above the component and add `<Description />` to storybook. I ran storybook and confirmed that the description appears.

vleague2 reviewed (commented):
Sorry for the delay in looking at this again! The code looks great. I have one more change request for a documentation update, and then we should be good to go!

vleague2 commented on libs/components/src/search/search.component.ts:
```diff
@@ -43,20 +47,34 @@ export class SearchComponent implements ControlValueAccessor, FocusableElement {
+
+  protected showResetButton = computed(() => this.isInputFocused() || this.isFormHovered());
+
+  @Input() disabled: boolean;
+  @Input() placeholder: string;
+  @Input() autocomplete: string;
```
Sorry about the merge conflicts, we just did a big migration to signal inputs for CL components. Can you put these back to how they are in `main`:
```suggestion
  readonly disabled = model<boolean>();
  readonly placeholder = input<string>();
  readonly autocomplete = input<string>();
```
And update their call sites back to how they are in `main`? 
- Line 97 in this file would go back to `this.disabled.set(isDisabled);`
- Line 20 in the template would call `placeholder()`
- Line 27 in the template would call `disabled()`
- Line 28 in the template would call `autocomplete()`

bensbits91 replied:
That makes sense. Thanks for pointing this out. Updated.

vleague2 reviewed (approved):
Thanks so much for your patience with all the change requests! I believe the UIF team will handle the QA, so hold off on merging until Robyn gives you the go-ahead!

bensbits91 commented:
QA passed in this PR: https://github.com/bitwarden/clients/pull/15749

Why? This original PR is sourced from my fork (because we started this before I worked here). That other PR was opened against a copy of the branch inside our Bitwarden repo to give QA access to the build artifacts. Now that QA passed, we'll merge this original PR because it has all the review comments and approval.
